File: two_pointers/triangle_validity/triangle_validity.py

def count_valid_triangles(edges: list[int]) -> int:  # LeetCode Q.611.
    edges.sort()
    total_triangles, total_edges = 0, len(edges)

    for edge_3_idx in range(total_edges - 1, -1, -1):
        edge_1_idx, edge_2_idx = 0, edge_3_idx - 1

        while edge_1_idx < edge_2_idx:
            if edges[edge_1_idx] + edges[edge_2_idx] > edges[edge_3_idx]:
                total_triangles += edge_2_idx - edge_1_idx
                edge_2_idx -= 1

            else:
                edge_1_idx += 1

    return total_triangles

File: two_pointers/triangle_validity/triangle_validity.cpp
#include <vector>
using namespace std;

int count_valid_triangles(vector<int> &edges)
{ // LeetCode Q.611.
    sort(edges.begin(), edges.end());

    int triangles = 0, min_edge_2_idx = 1;
    for (int edge_3_idx = 2; edge_3_idx < edges.size(); edge_3_idx++)
    {
        int edge_3_len = edges[edge_3_idx];
        while (edges[min_edge_2_idx - 1] + edges[min_edge_2_idx] <= edge_3_len)
        {
            min_edge_2_idx++;
            if (min_edge_2_idx >= edge_3_idx)
            {
                break;
            }
        }

        int min_edge_1_idx = min_edge_2_idx - 1;
        for (int edge_2_idx = min_edge_2_idx; edge_2_idx < edge_3_idx; edge_2_idx++)
        {
            int edge_2_len = edges[edge_2_idx];
            while (min_edge_1_idx > 0)
            {
                if (edges[min_edge_1_idx - 1] + edge_2_len <= edge_3_len)
                {
                    break;
                }
                min_edge_1_idx--;
            }

            triangles += edge_2_idx - min_edge_1_idx;
        }
    }

    return triangles;
}
